src/io/metadata.cpp

```diff
@@ -489,6 +521,41 @@ void Metadata::SetQuery(const data_size_t* query, data_size_t len) {
   #endif  // USE_CUDA
 }
 
+void Metadata::SetPosition(const data_size_t* positions, data_size_t len) {
+  std::lock_guard<std::mutex> lock(mutex_);
+  // save to nullptr
+  if (positions == nullptr || len == 0) {
+    positions_.clear();
+    num_positions_ = 0;
+    return;
+  }
+  if (num_data_ != len) {
+    Log::Fatal("Length of positions is not same with #data");
+  }
+  if (positions_.empty()) { positions_.resize(num_data_); }
+  num_positions_ = num_data_;
+
+  position_load_from_file_ = false;
+  #ifdef USE_CUDA
+  Log::Fatal("Positions in learning to rank is not supported in CUDA version yet.");
+  #endif  // USE_CUDA
+
+  #pragma omp parallel for schedule(static, 512) if (num_positions_ >= 1024)
+  for (data_size_t i = 0; i < num_positions_; ++i) {
+    positions_[i] = positions[i];
+  }
+
+  position_ids_.clear();
+  std::set<int> position_set;
+  for (int position : positions_) {
+    position_set.insert(position);
+  }
+  Log::Warning("position_set.size() = %ld", position_set.size());
```

Comment:
```suggestion
  Log::Debug("number of unique positions found = %ld", position_set.size());
```

This is going to result in a WARN-level log every time users use `positions`, right? Generally, warnings should be reserved for things that could cause issues and which users can control. I think this would be more appropriate as a DEBUG-level log.

In addition... will you consider my alternative text? Is it correct? I don't think putting literal code like `position_set.size()` in the log message will be very informative for users. Describing the data in real-world meaningful terms would be better.